streamer/src/nonblocking/rate_limiter.rs

```diff
@@ -0,0 +1,68 @@
+use {std::time::Duration, tokio::time::Instant};
+
+#[derive(Debug)]
+pub struct RateLimiter {
+    /// count of requests in an interval
+    pub(crate) count: u64,
+
+    /// Throttle start time
+    throttle_start_instant: Instant,
+    interval: Duration,
+    limit: u64,
+}
+
+/// A naive rate limiter, to be replaced by using governor
+impl RateLimiter {
+    pub fn new(limit: u64, interval: Duration) -> Self {
+        Self {
+            count: 0,
+            throttle_start_instant: Instant::now(),
+            interval,
+            limit,
+        }
+    }
+
+    /// Reset the counter and throttling start instant if needed.
+    pub fn reset_throttling_params_if_needed(&mut self) {
+        if Instant::now().duration_since(self.throttle_start_instant) > self.interval {
+            self.throttle_start_instant = Instant::now();
+            self.count = 0;
+        }
+    }
+    pub fn is_allowed(&mut self) -> bool {
+        self.reset_throttling_params_if_needed();
+        if self.count >= self.limit {
+            return false;
+        }
+
+        self.count = self.count.saturating_add(1);
```

Comment:
This seems a bad place to increment the count. It's expecting the user of the API to not call `is_allowed()` multiple times for the same connection/stream.

Reply:
Not sure I understand your concern. When a request is allowed to go through, we increment the count within that throttle window. I think even our stream throttling did the same thing so did governor. The difference in here is probably the counter is stored inside of the rate limiter.

Reply:
The function is named `is_allowed()`. Generally, such functions can be called any number of times yielding the same results. In this case, if it's called repeatedly, `self.count` will exceed `self.limit`, returning a different result.

We should move the count update out to a different function. Or, rename it and document it better. It seems error prone to me in its current form.

Reply:
I will update the documentation. This function is not supposed to be run multiple times for the same request.